File: src/main/java/sortingalgorithms/mergesort/java/MergeSort.java
package sortingalgorithms.mergesort.java;

import org.jetbrains.annotations.NotNull;
import sortingalgorithms.SortingAlgorithm;

public class MergeSort extends SortingAlgorithm {
    @Override
    public void sort(@NotNull int[] input) {
        int length = input.length;
        if (length < 2) return;
        int middle = length / 2;

        //1.Find the middle and divide the array O(log)
        int[] leftArray = new int[middle];
        int[] rightArray = new int[length - middle];

        for (int i = 0; i < middle; i++) {
            leftArray[i] = input[i];
        }

        for (int j = middle; j < length; j++) {
            rightArray[j - middle] = input[j];
        }

        //2. Sort the both arrays
        sort(leftArray);
        sort(rightArray);

        //3. Final Step is to merge
        merge(leftArray, rightArray, input);
    }

    private void merge(int[] leftArray, int[] rightArray, int[] input) {
        int i = 0, j = 0, k = 0;

        while (i < leftArray.length && j < rightArray.length) {
            if (leftArray[i] <= rightArray[j]) {
                input[k++] = leftArray[i++];
            } else {
                input[k++] = rightArray[j++];
            }
        }

        while (i < leftArray.length) {
            input[k++] = leftArray[i++];
        }

        while (j < rightArray.length) {
            input[k++] = rightArray[j++];
        }
    }
}

File: src/main/java/sortingalgorithms/mergesort/kotlin/MergeSort.kt
package sortingalgorithms.mergesort.kotlin

import sortingalgorithms.SortingAlgorithm

class MergeSort : SortingAlgorithm() {
    override fun sort(input: Array<Int>) {
        val length = input.size
        if (length < 2) return //Base condition
        //1.Find the middle and divide the array O(log)
        var middle = length / 2
        var leftArray = Array(middle) { 0 } // 0 to middle
        var rightArray = Array(length - middle) { 0 } // middle to end of the array (length)

        for (i in 0 until middle) {
            leftArray[i] = input[i]
        }

        for (j in middle until length) {
            //j - middle : make sure to get the right index
            rightArray[j - middle] = input[j]
        }

        //2. We sort both the arrays Recursively
        sort(leftArray)
        sort(rightArray)

        //3. Last step let's merge it
        merge(leftArray, rightArray, input)
    }

    private fun merge(leftArray: Array<Int>, rightArray: Array<Int>, inputData: Array<Int>) {
        var i = 0
        var j = 0
        var k = 0
        // when both left and right array sizes are equal
        while (i < leftArray.size && j < rightArray.size) {
            if (leftArray[i] <= rightArray[j]) {
                inputData[k++] = leftArray[i++]
            } else {
                inputData[k++] = rightArray[j++]
            }
        }

        while (i < leftArray.size) {
            inputData[k++] = leftArray[i++]
        }

        while (j < rightArray.size) {
            inputData[k++] = rightArray[j++]
        }
    }
}